<a href="https://colab.research.google.com/github/2203A51455N/DATA/blob/main/Week_1_Monday_28_07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze student marks data to classify scores, identify patterns, and provide data-driven insights for improving academic outcomes.

## Data preparation & classification

### Subtask:
Create a list of 10 student marks, define a function to classify each score into "Fail", "Borderline", or "Pass", and convert the results into a pandas categorical column.


**Reasoning**:
Create a list of student marks, define a function to classify scores, apply the function to the list, and create a pandas categorical Series from the results.



In [ ]:
import pandas as pd

# 1. Create a Python list named student_marks
student_marks = [45, 62, 78, 30, 55, 88, 69, 92, 49, 70]

# 2. Define a Python function named classify_score
def classify_score(mark):
    if mark < 50:
        return "Fail"
    elif 50 <= mark < 70:
        return "Borderline"
    else:
        return "Pass"

# 3. Apply the classify_score function to each mark in the student_marks list
classifications = [classify_score(mark) for mark in student_marks]

# 4. Create a pandas Series from the list of classifications and convert to categorical
classified_marks = pd.Series(classifications, dtype='category')

# Display the classified marks
display(classified_marks)

0          Fail
1    Borderline
2          Pass
3          Fail
4    Borderline
5          Pass
6    Borderline
7          Pass
8          Fail
9          Pass
dtype: category
Categories (3, object): ['Borderline', 'Fail', 'Pass']

## Data exploration

### Subtask:
Display the frequency count and percentage of students in each category, and show the data type and structure of the classified data.


**Reasoning**:
Calculate and display the frequency count and percentage of each category, and display the data type and structure of the `classified_marks` Series as requested by the instructions.



In [ ]:
# Calculate and display the frequency count of each category
frequency_count = classified_marks.value_counts()
print("Frequency Count:")
display(frequency_count)

# Calculate and display the percentage of students in each category
percentage = classified_marks.value_counts(normalize=True) * 100
print("\nPercentage:")
display(percentage)

# Display the data type of the classified_marks Series
print("\nData Type:")
display(classified_marks.dtype)

# Display the structure (head) of the classified_marks Series
print("\nStructure (Head):")
display(classified_marks.head())

Frequency Count:


Pass          4
Borderline    3
Fail          3
Name: count, dtype: int64


Percentage:


Pass          40.0
Borderline    30.0
Fail          30.0
Name: proportion, dtype: float64


Data Type:


CategoricalDtype(categories=['Borderline', 'Fail', 'Pass'], ordered=False, categories_dtype=object)


Structure (Head):


0          Fail
1    Borderline
2          Pass
3          Fail
4    Borderline
dtype: category
Categories (3, object): ['Borderline', 'Fail', 'Pass']

## Statistical insight

### Subtask:
Compute and display the mean, median, and standard deviation of the marks, and the number of students at risk ("Fail" or "Borderline").


**Reasoning**:
Calculate and print the mean, median, and standard deviation of the student marks and count and print the number of students at risk.



In [ ]:
import numpy as np

# Calculate the mean, median, and standard deviation of the student_marks list
mean_marks = np.mean(student_marks)
median_marks = np.median(student_marks)
std_dev_marks = np.std(student_marks)

# Print the calculated values
print(f"Mean of student marks: {mean_marks:.2f}")
print(f"Median of student marks: {median_marks:.2f}")
print(f"Standard deviation of student marks: {std_dev_marks:.2f}")

# Count the number of students classified as "Fail" or "Borderline"
at_risk_students = classified_marks[
    (classified_marks == "Fail") | (classified_marks == "Borderline")
].count()

# Print the count of students at risk
print(f"Number of students at risk (Fail or Borderline): {at_risk_students}")

Mean of student marks: 63.80
Median of student marks: 65.50
Standard deviation of student marks: 18.56
Number of students at risk (Fail or Borderline): 6


## Summary:

### Data Analysis Key Findings

*   Student marks were classified into "Fail" (\$< 50\$), "Borderline" (\$50 \leq \text{mark} < 70\$), and "Pass" (\$ \geq 70\$).
*   Out of 10 students, 40% passed, 30% were borderline, and 30% failed.
*   The mean student mark is approximately 63.80, with a median of 65.50.
*   The standard deviation of the marks is approximately 18.56, suggesting a moderate spread in performance.
*   6 students are identified as being at risk, classified as either "Fail" or "Borderline".

### Insights or Next Steps

*   Focus interventions on the 6 students identified as at risk ("Fail" or "Borderline") to improve academic outcomes.
*   Analyze the distribution of marks further to identify potential areas for curriculum or teaching method adjustments.
